ui/lib/apps/QueryEditor/index.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { useState, useCallback, useRef } from 'react'
+import cx from 'classnames'
+import { Root, Card } from '@lib/components'
+import Split from 'react-split'
+import { Button, Modal, Space, Typography } from 'antd'
+import {
+  CaretRightOutlined,
+  LoadingOutlined,
+  WarningOutlined,
+  CheckOutlined,
+} from '@ant-design/icons'
+
+import Editor from './Editor'
+import ResultTable from './ResultTable'
+
+import styles from './index.module.less'
+import client, { QueryeditorRunResponse } from '@lib/client'
+import ReactAce from 'react-ace/lib/ace'
+import { getValueFormat } from '@baurine/grafana-value-formats'
+
+const MAX_DISPLAY_ROWS = 1000
+
+function App() {
+  const [results, setResults] = useState<QueryeditorRunResponse | undefined>()
+  const [isRunning, setRunning] = useState(false)
+  const editor = useRef<ReactAce>(null)
+
+  const isResultsEmpty =
+    !results ||
+    (!results.error_msg && (!results.column_names?.length || !results.rows))
+
+  const handleRun = useCallback(async () => {
+    setRunning(true)
+    setResults(undefined)
+    try {
+      const resp = await client.getInstance().queryEditorRun({
+        max_rows: MAX_DISPLAY_ROWS,
+        statements: editor.current?.editor.getValue(),
+      })
+      setResults(resp.data)
+    } catch (ex) {
+      Modal.error({
+        content: ex.message,
+      })
+    }
+    setRunning(false)
+    editor.current?.editor.focus()
+  }, [])
```

Comment:
Not sure whether we can use `useClientRequest()` to simplify the code like this:

```
const {data, isLoading, error, sendRequest: handleRun} = useClientRequest((cancelToken) => 
    client.getInstance().queryEditorRun({
        max_rows: MAX_DISPLAY_ROWS,
        statements: editor.current?.editor.getValue(),
    }, {cancelToken}), 
    { immediate: false, afterRequest: () => editor.current?.editor.focus() }
)
```

Then we don't need `results` and `isRunning` anymore.

Reply:
Let me try it.

Reply:
Unlike queries, for operational errors mostly there will be a modal dialog to stop further user actions. Displaying modal dialog using this pattern can be pretty hard.